src/main/java/com/hierynomus/smbj/share/File.java

```diff
@@ -114,14 +115,22 @@ public int read(byte[] buffer, long fileOffset) throws IOException {
      * @param fileOffset The offset, in bytes, into the file from which the data should be read
      * @param offset the start offset in the buffer at which to write data
      * @param length the maximum number of bytes to read
-     * @return the actual number of bytes that were read
+     * @return the actual number of bytes that were read; or -1 if the end of the file was reached
      */
     public int read(byte[] buffer, long fileOffset, int offset, int length) throws IOException {
         SMB2ReadResponse response = share.read(fileId, fileOffset, length);
-        byte[] data = response.getData();
-        int bytesRead = Math.min(length, data.length);
-        System.arraycopy(data, 0, buffer, offset, bytesRead);
-        return bytesRead;
+        if (response.getHeader().getStatus() == NtStatus.STATUS_END_OF_FILE) {
+            return -1;
```

Comment:
See https://msdn.microsoft.com/en-us/library/cc246729.aspx?f=255&MSPPError=-2147217396 for the details.
SMBJ sets MinimumCount in the read request to 1. So we should only be getting END_OF_FILE if there is no data left. In any case, SMBJ currently doesn't parse the data even if there is any unless STATUS_SUCCESS is returned. See com/hierynomus/mssmb2/messages/SMB2ReadResponse.java:38